### PR TITLE
Flush act() only on exiting outermost call

### PR DESCRIPTION
**Pull request by threepointone**
Just starting this PR up to discuss the changes. The actual code changes aren't major. The big (breaking?) change is that nested acts won't work anymore.

### CODE REVIEW

**Comment by acdlite** [comment]
Will wait for https://github.com/facebook/react/pull/15591 to land before reviewing (or I can review now if you base this one on top of that one)

File: packages/react-dom/src/test-utils/ReactTestUtilsAct.js [hunk 1]
```diff
@@ -119,6 +118,13 @@ function act(callback: () => Thenable) {
         called = true;
         result.then(
           () => {
+            if (actingUpdatesScopeDepth !== 1) {
```
Comment (acdlite) [inline packages/react-dom/src/test-utils/ReactTestUtilsAct.js]: Is there any scenario where `actingUpdatesScopeDepth` would decrement twice, and this would be less than 1? Maybe should change to `actingUpdatesScopeDepth < 1` just in case.

File: packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js [hunk 2]
```diff
@@ -285,14 +285,13 @@ describe('ReactTestUtils.act()', () => {
         return ctr;
       }
       const el = document.createElement('div');
+      act(() => {
```
Comment (acdlite) [inline packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js]: Why did you move this to a separate `act`? I would expect moving the assertion outside `act` is sufficient.

File: packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js [hunk 3]
```diff
@@ -371,18 +370,23 @@ describe('ReactTestUtils.act()', () => {
...
+      await act(async () => {}); // #just-react-things
 
       expect(div.innerHTML).toBe('1');
       expect(ctr).toBe(2);
+
+      // this may seem odd, but it matches browser behaviour -
+      // a flash of "0" followed by "1"
+
+      // now you probably wouldn't write a test like this for your app
+      // you'd await act(async () => { ReactDOM.render(<App/> })
```
Comment (acdlite) [inline packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js]: So then why did you write it this way?

File: packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js [hunk 4]
```diff
@@ -371,18 +370,23 @@ describe('ReactTestUtils.act()', () => {
...
-        expect(ctr).toBe(1);
+      act(() => {
+        ReactDOM.render(<App callback={() => ctr++} />, div);
       });
-      // this may seem odd, but it matches user behaviour -
-      // a flash of "0" followed by "1"
+      expect(div.innerHTML).toBe('0');
+      expect(ctr).toBe(1);
+
+      await act(async () => {}); // #just-react-things
```
Comment (acdlite) [inline packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js]: Why did you remove the `await` from the earlier `act`?

File: packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js (hunk 2 above)
Comment (threepointone) [reply packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js]: the effect wouldn't flush effects until the act exits, only after which the timeout in the component would start, failing the assertion. hence, 2 act()s.

File: packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js (hunk 4 above)
Comment (threepointone) [reply packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js]: I wanted to specifically test the intermediate sync state. this test is named badly (and frankly on looking at it now, it doesn't seem super useful).

File: packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js (hunk 3 above)
Comment (threepointone) [reply packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js]: like mentioned, just wanted to make sure the intermediate state was what I'd expected.

File: packages/react-dom/src/test-utils/ReactTestUtilsAct.js (hunk 1 above)
Comment (threepointone) [reply packages/react-dom/src/test-utils/ReactTestUtilsAct.js]: rather, actingUpdatesScopeDepth > 1

File: packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js (hunk 4 above)
Comment (acdlite) [reply packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js]: It's not clear to me at all what behavior you're actually trying to test. More comments could help.

File: packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js (hunk 2 above)
Comment (acdlite) [reply packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js]: Why did you remove the `await` then?

**Comment by threepointone** [comment]
so many merge conflicts, so I redid it here https://github.com/facebook/react/pull/15682
